Added more recording options

devalentineomonya opened the pull request:
### Pull Request: Refactor and Extend `record` Subcommand in Caelestia CLI

This pull request refactors and extends the `record` subcommand in the **Caelestia CLI**, improving flexibility, reliability, and user experience for screen recording. It introduces granular video and audio options, robust device handling, enhanced notifications, and cleaner code formatting.

---

#### **Enhancements to `record` subcommand**

* **Granular recording modes**:

  * **Video modes**: `fullscreen`, `region`, `window`
  * **Audio modes**: `none`, `mic`, `system`, `combined`
    These replace previous simpler options, allowing more flexible recording scenarios, such as tutorials, presentations, or system-only captures.

* **Improved audio device selection**:

  * Added a `get_audio_device` method that validates device availability and provides fallbacks, increasing reliability when recording audio.
  * Supports distinct selection for microphone, system audio, or both.

* **Backward compatibility**:

  * Legacy flags like `--sound` are still supported for users upgrading from older versions.

---

#### **User Experience Improvements**

* **Status and control flags**:

  * `--status` to check ongoing recording
  * `--stop` to terminate a recording gracefully

* **Enhanced notifications**:

  * Non-blocking notifications on recording completion
  * Clear warnings if recording file is missing

* **Error handling improvements**:

  * Friendly warnings for window selection failures or file operation errors

---

#### **Code Quality and Maintainability**

* Refactored argument parsing for readability and maintainability
* Improved docstrings and comments for clarity
* Consistent formatting across subcommands

---

#### **Usage Example Screenshot**

<img width="431" height="461" alt="Screen recording options example" src="https://github.com/user-attachments/assets/baa0b7de-0e19-48ab-9974-5df07b2f4028" />

---

These changes make the `record` subcommand more robust, flexible, and suitable for tutorials or professional recording workflows.

The changes work well with the **shell fork** on my repo: [Caelestia Shell](https://github.com/devalentineomonya/caelestia-shell) and can be forked and compiled directly.

devalentineomonya committed:
My local changes

devalentineomonya committed:
Added my screen recorder config

devalentineomonya committed:
Updated audio options

github-actions[bot] committed:
[CI] chore: update flake

devalentineomonya committed:
Fixed window option

devalentineomonya committed:
Fixed window option

github-actions[bot] committed:
[CI] chore: update flake

github-actions[bot] committed:
[CI] chore: update flake

github-actions[bot] committed:
[CI] chore: update flake

github-actions[bot] committed:
[CI] chore: update flake

github-actions[bot] committed:
[CI] chore: update flake

github-actions[bot] committed:
[CI] chore: update flake

github-actions[bot] committed:
[CI] chore: update flake

github-actions[bot] committed:
[CI] chore: update flake

github-actions[bot] committed:
[CI] chore: update flake

MateusAquino commented:
I'm interested on this feature, it also closes https://github.com/caelestia-dots/shell/issues/945, but not sure if it's a great idea to replace the default "-s" argument from "sound" to "stop"; not to mention this change requires an existing "Combined" sink to work, which will not exist by default:

<img width="1209" height="674" alt="image" src="https://github.com/user-attachments/assets/8f78af2f-0f96-4b43-b1b7-650ed7efacbe" />

<br/><br/>
Also note that recording options overflows when no recordings exists + the list is collapsed:

<img width="575" height="421" alt="image" src="https://github.com/user-attachments/assets/3a922b5d-39a9-49d7-8415-cf8cce559a43" />

devalentineomonya committed:
Update src/caelestia/subcommands/record.py

Co-authored-by: Mateus <mateusaqb@gmail.com>

github-actions[bot] committed:
[CI] chore: update flake

github-actions[bot] committed:
[CI] chore: update flake

github-actions[bot] committed:
[CI] chore: update flake

github-actions[bot] committed:
[CI] chore: update flake

github-actions[bot] committed:
[CI] chore: update flake

github-actions[bot] committed:
[CI] chore: update flake

soramanew committed:
theme: update discord theme template

Fixes caelestia-dots/caelestia#206

github-actions[bot] committed:
[CI] chore: update flake

github-actions[bot] committed:
[CI] chore: update flake

mj0x0 committed:
theme: ensure atomic writes for configuration files (#76)

* fix: ensure atomic writes for configuration files

* requested changes

github-actions[bot] committed:
[CI] chore: update flake

github-actions[bot] committed:
[CI] chore: update flake

github-actions[bot] committed:
[CI] chore: update flake

github-actions[bot] committed:
[CI] chore: update flake

github-actions[bot] committed:
[CI] chore: update flake

Kalagmitan committed:
theme: inject mode into user templates (#77)

* theme: apply_user_templates accepts mode.

* Some themes like those in NvChad require a mode to be supplied to
work. Added a minimal change that makes apply_user_templates accept a
mode parameter and replaces any {{ mode }} placeholder in a file with
the actual mode.

* theme: mode replace integrated with gen_replace_dynamic

+ Moved the {{ mode }} replacement logic to the gen_replace_dynamic
function.

* refactor: adjusted comment

github-actions[bot] committed:
[CI] chore: update flake

github-actions[bot] committed:
[CI] chore: update flake

github-actions[bot] committed:
[CI] chore: update flake

github-actions[bot] committed:
[CI] chore: update flake

github-actions[bot] committed:
[CI] chore: update flake

github-actions[bot] committed:
[CI] chore: update flake

github-actions[bot] committed:
[CI] chore: update flake

github-actions[bot] committed:
[CI] chore: update flake

github-actions[bot] committed:
[CI] chore: update flake

github-actions[bot] committed:
[CI] chore: update flake

github-actions[bot] committed:
[CI] chore: update flake

chea-vuthearith committed:
record: add arg to copy screen recording to clipboard (#83)

soramanew committed:
fix: xxx_paletteKeyColor rename in myc 3.0.0

Fixes caelestia-dots/shell#1112

AteebXYZ committed:
fix: apply mode substitution after dynamic template fill (#84)

nicklany01 committed:
nix: fix todoist command replacement in default.nix (#86)

Target the command list specifically ["todoist"] instead of a global
replace to avoid breaking the configuration key structure.

github-actions[bot] committed:
[CI] chore: update flake

github-actions[bot] committed:
[CI] chore: update flake

soramanew committed:
fix: compat for python-materialyoucolor < 3.0.0

PixelKhaos committed:
feat: thunar & papirus-folders theming + new schemes (#80)

* feat: GTK app theming system

- Implemented custom.css import for user-managed app themes
- process_app_themes() to dynamically update colors in imported CSS files
- Inline comment markers for color replacement (e.g, /* accent-color */)
- Papirus icon color syncing with weighted hue/saturation algorithm

This allows users to create modular app themes that automatically update when the scheme/wallpaper changes

Example usage:
  .app .element { color: #24BD5C; /* accent-color */ }
  .app .element:hover { background: rgba(36, 189, 92, 0.15); /* accent-color with 15% opacity */ }

* feat: atomic theme changes with locking and mode-specific CSS

- Implemented locking to prevent concurrent theme changes
- Added mode-light/mode-dark CSS markers for dynamic property reordering
- Made terminal writes and Papirus sync non-blocking to prevent hangs
- Only save scheme.json after successful theme application

Fixes race conditions during rapid theme switching and ensures Shell and GTK apps scheme stay in sync.

* theme: added to color mapping for custom theming, new schemes

* theme: quick fixes, cleanup

* theme: include thunar.css as template, with new theming system

* theme: modified GTK theming approach

- Dropped comment targeted theming in favor for existing {{  }} replacement
- [app].css.template file created for customization, bypassing built in default if present
- Handling *.template for added templates to be parsed and added to import

* theme: fixes for  thunar.css

* theme: remove .template file use

* theme: path button color adjustment, non-active hover

* fixes & cleanup

* thunar css fixes

* more css fixes

* format

* fix tab vert spacing

---------

Co-authored-by: 2 * r + 2 * t <61896496+soramanew@users.noreply.github.com>

soramanew committed:
feat: add dynamic hard flavour

Someone requested idk who tho

github-actions[bot] committed:
[CI] chore: update flake

Unrectified committed:
feat: scheme support for Pandora Minecraft launcher (#87)

* added pandora theme template json

* feat: add support for Pandora theme integration

soramanew committed:
fix: dynamic scheme import

github-actions[bot] committed:
[CI] chore: update flake

github-actions[bot] committed:
[CI] chore: update flake

Unrectified committed:
feat: add GIF files support as wallpaper (#88)

* feat: add GIF files support as wallpaper (not animated tho)

* tweak: simplifying variable use

* fix: git is good but damn it's annoying

* fix

---------

Co-authored-by: 2 * r + 2 * t <61896496+soramanew@users.noreply.github.com>

soramanew committed:
fix: missing colon

github-actions[bot] committed:
[CI] chore: update flake

github-actions[bot] committed:
[CI] chore: update flake

github-actions[bot] committed:
[CI] chore: update flake

github-actions[bot] committed:
[CI] chore: update flake

soramanew committed:
fix: dynamic scheme import <3.0.0 compat

github-actions[bot] committed:
[CI] chore: update flake

github-actions[bot] committed:
[CI] chore: update flake

xxeisenberg committed:
fix: wall not Path type (#89)

github-actions[bot] committed:
[CI] chore: update flake

github-actions[bot] committed:
[CI] chore: update flake

github-actions[bot] committed:
[CI] chore: update flake

soramanew committed:
fix: format

soramanew committed:
feat: switch to qtengine

Kalagmitan committed:
refactor: enforce stricter type hints (#91)

LSP was screaming at me so I decided to just address it to get it off my
screen.

+ Fixed the type hints
:= Modified and added type hints for certain functions and variables in
most of the files in the utils/ folder (and some in the subcommands/
folder) for clarity and so pyright's type checker wouldn't cry.
:+ To resolve certain type issues, I had to add a bit more tiny
additional code such as, additional checks if a variable is None, a tiny
class in utils/material/generator.py to resolve the constructor usage
mismatch between what the DynamicScheme accepts and what the code
actually passes, and etc.
- Renamed certain functions and variables for clarity and also for some
to not collide with pre-existing definitions from well-known library
imports.
+ PIL has reorganized their code a bit, so the code is made to reflect
their new definitions.
= Reorganized the single import statement for "colourfulness" in
utils/wallpaper.py to be close to the top.
(I think that's it)

Side Effects?:
Everything should work the same as no logic change was done whatsover
(unless we consider the added if statements for type checking as a logic
change). I've tested it, everything seems to be in urdir.

SamiyelF committed:
nix: fix qtct sub (#98)

github-actions[bot] committed:
[CI] chore: update flake

github-actions[bot] committed:
[CI] chore: update flake

github-actions[bot] committed:
[CI] chore: update flake

github-actions[bot] committed:
[CI] chore: update flake

AteebXYZ committed:
readme: document user templates (#99)

* document user_templates in README

* move user_template documentation to Usage

* format

---------

Co-authored-by: 2 * r + 2 * t <61896496+soramanew@users.noreply.github.com>

github-actions[bot] committed:
[CI] chore: update flake

github-actions[bot] committed:
[CI] chore: update flake

github-actions[bot] committed:
[CI] chore: update flake

github-actions[bot] committed:
[CI] chore: update flake

github-actions[bot] committed:
[CI] chore: update flake

github-actions[bot] committed:
[CI] chore: update flake

github-actions[bot] committed:
[CI] chore: update flake

github-actions[bot] committed:
[CI] chore: update flake

github-actions[bot] committed:
[CI] chore: update flake

github-actions[bot] committed:
[CI] chore: update flake

github-actions[bot] committed:
[CI] chore: update flake

github-actions[bot] committed:
[CI] chore: update flake

github-actions[bot] committed:
[CI] chore: update flake

github-actions[bot] committed:
[CI] chore: update flake

github-actions[bot] committed:
[CI] chore: update flake

github-actions[bot] committed:
[CI] chore: update flake

soramanew committed:
feat: add caelestia scheme

Closes #100

github-actions[bot] committed:
[CI] chore: update flake

yukazakiri committed:
theme: add zed editor theme support (#102)

- Add Zed theme template (zed.json) with full Material You syntax
  highlighting, UI colors, and terminal ANSI palette
- Apply theme via file overwrite instead of restarting Zed, so
  Zed's file watcher picks up changes automatically
- Resolve symlinks in apply_zed() only before writing, since
  Zed's file watcher does not detect changes through symlinks
  (other themed apps that rely on symlinks are unaffected)

soramanew committed:
fix: remove troublesome resizer rules

yukazakiri committed:
theme: add live theming for chromium-based browsers (#103)

* theme: add live theming for chromium-based browsers

Writes the surface colour as an RGB triplet to chromium.theme and
applies it as a managed browser policy for chromium, brave, and
google-chrome-stable using the --refresh-platform-policy flag
(introduced in Chrome 142+).

Also fixes write_file to set 0644 permissions so browser processes
can read the policy files.

* feat: use sudo tee not chmod

* chore: update readme

---------

Co-authored-by: 2 * r + 2 * t <61896496+soramanew@users.noreply.github.com>

soramanew committed:
docs: add missing theme.enable* opts to example conf

github-actions[bot] committed:
[CI] chore: update flake

github-actions[bot] committed:
[CI] chore: update flake

devalentineomonya committed:
UPdated recording util combined sink

devalentineomonya committed:
Updated screenrecorder to re-encord audio to aac

devalentineomonya committed:
Added logging for screenshot
